### ISSUE TITLE
Add a way to generate symbolic information for Windows builds

### ISSUE THREAD

**Issue by neitsa** [CleverRaven/Cataclysm-DDA#29248]
As of now the Windows build stack traces are mostly inoperable, unless someone pulls out the right binary and looks directly at the right offsets in the faulting binary.

This discussion revolves around having symbolic information in program databases (as known as PDB) for C:DDA Windows builds to alleviate this problem and get meaningful stack traces.

# What is symbolic information?

Symbolic information can be generated for debug or release versions of a binary and contain various level of information depending on the need such as:

* Global variables
* Local variables
* Function names and the addresses of their entry points
* Frame pointer omission (FPO) records
* Source-line numbers

Symbolic information is mainly useful for debugging but also for generating a meaningful crash report.

This symbolic information can be integrated inside the binary or kept outside the binary in a dedicated file. In the latter case this symbolic information - generated by the linker while linking a binary - is enclosed in an external file, in what is named a "program database" which has the `.pdb` extension.

# Why Symbolic information is needed for C:DDA?

Symbolic information would be useful to generate a valid and meaningful stack trace on Windows systems. In this case we would only need the following symbolic information:

* Function names and the addresses of their entry points
* Source-line numbers

As of now the binaries generated by the Linux Mingw compiler only provide DWARF symbolic information which is not understood by Windows systems.

On the other hand, once Microsoft symbolic information is available, there are multiple system APIs (i.e there's no need  of third parties application or development kit) that can read this symbolic information and generate a meaningful stack trace  and also a "dump" file (also know as "core" file).
 
As an example here is a stack trace **without** symbolic information:

```
STACK TRACE:
@0x54C804[cataclysm-tiles.exe+0x14C804]
@0x54CBA5[cataclysm-tiles.exe+0x14CBA5]
SMPEG_error+0x3DAF3@0xBA2655[cataclysm-tiles.exe+0x7A2655]
UnhandledExceptionFilter+0x1A0@0x77815F10[KERNELBASE.dll+0x195F10]
RtlCaptureStackContext+0x1E521@0x77BBDA71[ntdll.dll+0x9DA71]
RtlAreBitsSet+0x58@0x77B84198[ntdll.dll+0x64198]
```

Here's the stack trace for the same crash **with** *full* symbolic information:

```
00 (Inline Function) : Cataclysm!std::basic_string<char,std::char_traits<char>,std::allocator<char> >::_Check_offset [d:\programming\microsoft visual studio 14.0\vc\include\xstring @ 2271] 
01 00007ff6`0b1ff8c8 : Cataclysm!std::basic_string<char,std::char_traits<char>,std::allocator<char> >::assign+0x15 [d:\programming\microsoft visual studio 14.0\vc\include\xstring @ 1150] 
02 (Inline Function) : Cataclysm!std::basic_string<char,std::char_traits<char>,std::allocator<char> >::{ctor}+0x20 [d:\programming\microsoft visual studio 14.0\vc\include\xstring @ 817] 
03 00007ff6`0bb5d959 : Cataclysm!std::basic_string<char,std::char_traits<char>,std::allocator<char> >::substr+0x28 [d:\programming\microsoft visual studio 14.0\vc\include\xstring @ 2089] 
04 00007ff6`0bb8a68c : Cataclysm!vehicle::remove_carried_vehicle+0x569 [cataclysm-dda\src\vehicle.cpp @ 1676] 
05 00007ff6`0bb82cdf : Cataclysm!vehicle::use_bike_rack+0x6cc [cataclysm-dda\src\vehicle_use.cpp @ 1347] 
06 00007ff6`0b9a9002 : Cataclysm!vehicle::interact_with+0xecf [cataclysm-dda\src\vehicle_use.cpp @ 1482] 
07 00007ff6`0b477491 : Cataclysm!Pickup::pick_up+0xc2 [cataclysm-dda\src\pickup.cpp @ 365] 
08 00007ff6`0b477ae9 : Cataclysm!game::examine+0x181 [cataclysm-dda\src\game.cpp @ 5962] 
09 00007ff6`0b4cd5a6 : Cataclysm!game::examine+0x199 [cataclysm-dda\src\game.cpp @ 5879] 
0a 00007ff6`0b46fbb4 : Cataclysm!game::handle_action+0xfc6 [cataclysm-dda\src\handle_action.cpp @ 1471] 
0b 00007ff6`0b1f3f3c : Cataclysm!game::do_turn+0x714 [cataclysm-dda\src\game.cpp @ 1402] 
0c 00007ff6`0b161232 : Cataclysm!SDL_main+0x117c [cataclysm-dda\src\main.cpp @ 682] 
0d 00007ff6`0bbe9663 : Cataclysm!main_getcmdline+0xf2 [c:\projects\sdl\src\main\windows\sdl_windows_main.c @ 177] 
0e (Inline Function) : Cataclysm!invoke_main+0x21 [f:\dd\vctools\crt\vcstartup\src\startup\exe_common.inl @ 94] 
0f 00007ffc`a85c81f4 : Cataclysm!__scrt_common_main_seh+0x117 [f:\dd\vctools\crt\vcstartup\src\startup\exe_common.inl @ 253] 
10 00007ffc`a9fca251 : KERNEL32!BaseThreadInitThunk+0x14
11 00000000`00000000 : ntdll!RtlUserThreadStart+0x21
```

# Problematics

## Build Chain

The first problem is that, to generate symbolic information which is understood natively by a Windows system, there's a need to change the current build chain (compiler and linker) to programs that follow the Microsoft ABI.

As of now the Windows builds are generated by Linux Mingw which is not capable of outputting Microsoft native symbolic information (it outputs DWARF symbolic information which is native to Linux).

### Possible Solutions

* Use a Windows Virtual Machine (VM) with the free [Visual Studio Build Tools](https://devblogs.microsoft.com/cppblog/visual-studio-build-tools-now-include-the-vs2017-and-vs2015-msvc-toolsets/)  (that is, only the build tools and no IDE.)
    - This obviously requires a builder than can run a VM...
    - Needs a valid Windows license...
    
* Pull out the resulting binary (`*.exe`) and its `*.pdb` from the Appveyor builder.
    - This requires an [external storage](https://www.appveyor.com/docs/deployment/) (FTP, Azure, Bintray or Github) where Appveyor can upload the resulting binaries.
    - This also requires to have a way to differentiate between a simple PR check build and a daily build (maybe [through a tag?](https://www.appveyor.com/docs/branches/#conditional-build-configuration)).

* Use the latest Clang version (8.0 as of now) as the build chain. 
    - [It can output](http://blog.llvm.org/2017/08/llvm-on-windows-now-supports-pdb-debug.html) the correct **native** Microsoft symbolic information.
    - It *seems* to be [usable on Linux too](http://nosubstance.me/post/coding-windows-cpp-on-linux/)!
    - See also [clang-cl](https://clang.llvm.org/docs/UsersManual.html#clang-cl) which uses the same switches than the MS compiler.
    - Can be used with [CMake](https://github.com/ProdriveTechnologies/clang-msvc-sdk).
    
* Convert DWARF information to PDB
    - By using [cv2pdb](https://github.com/rainers/cv2pdb)
    - Seems to generate some errors though.
    
* Convert the current Mingw/GCC build chain to [MinGw/LLVM](https://github.com/mstorsjo/llvm-mingw)
    - It also offers [PDB generation](https://github.com/mstorsjo/llvm-mingw#pdb-support).
    - This is a glorified wrapper around clang though.

## Symbolic information Size

Supposing that we have now a build chain that can output native Microsoft symbolic information, we need to distribute the symbolic information to get valid stack traces.

The main problem is that symbolic information can take a **lot** of room.

Here are various sizes of the symbolic information for C:DDA for various builds and various PDB options:

- [**debug** + shared libs] Full symbolic information: 370 MB (yes that's Megabytes)
- [**release** + shared libs] Full symbolic information: 182 MB

### Solution 

The stripped PDB (using the [`/PDBSTRIPPED`](https://docs.microsoft.com/en-us/cpp/build/reference/pdbstripped-strip-private-symbols?view=vs-2019) linker switch)
allows to output a secondary PDB file without private symbols. Fortunately these public symbols retain at least function names and addresses (but not function lines...).

- [**release** build + shared libs] Stripped symbolic information: 30 MB

The question is: would it be acceptable to add around 30 MB to the release archive?

[TODO] follow up on having an external dumper.

**Comment by ZhilkinSerg** [comment]
> The question is: would it be acceptable to add around 30 MB to the release archive?

We already add ~25 Mb of debug info to Windows builds - if that would just be replaced with ~30 Mb PDB, there won't be much difference in download size.

**Comment by kevingranade** [comment]
Building with clang or using cv2pdb are good options. I'm not excited about growing a Microsoft based dependency in the build infrastructure, but if someone donated access to a build system I don't think I'd turn it down.
The appveyor thing would also be fine probably, but seems like more hassle than the alternatives.

I don't think an extra 30MB would be unreasonable considering the benefits.

**Comment by ZhilkinSerg** [comment]
cv2pdb results (crash from debug menu):

![image](https://user-images.githubusercontent.com/16213433/55495811-6729c180-5646-11e9-9fc4-c116c86a82ea.png)

### original:
```
CRASH LOG FILE: config/crash.log
VERSION: 0.D-1370-gb52c60f
TYPE: Signal
MESSAGE: SIGSEGV: Segmentation fault
STACK TRACE:
    @0x558475[cataclysm-tiles.exe+0x158475]
    @0x559022[cataclysm-tiles.exe+0x159022]
    raise+0x200@0x7FFF3B6BAD70[msvcrt.dll+0x2AD70]
    @0x6755E1[cataclysm-tiles.exe+0x2755E1]
    @0x6A39E0[cataclysm-tiles.exe+0x2A39E0]
    @0x686557[cataclysm-tiles.exe+0x286557]
    IMG_LoadWEBP_RW+0x4DEEF9@0x140A8D9[cataclysm-tiles.exe+0x100A8D9]
    @0x4013ED[cataclysm-tiles.exe+0x13ED]
    @0x4014FB[cataclysm-tiles.exe+0x14FB]
    BaseThreadInitThunk+0x14@0x7FFF39888364[KERNEL32.DLL+0x8364]
    RtlUserThreadStart+0x21@0x7FFF3B8D70D1[ntdll.dll+0x670D1]
```

### pdb (after cv2pdb):

```
CRASH LOG FILE: config/crash.log
VERSION: 0.D-1370-gb52c60f
TYPE: Signal
MESSAGE: SIGSEGV: Segmentation fault
STACK TRACE:
    mainCRTStartup+0x156F65@0x558475[cataclysm-tiles.exe+0x158475]
    mainCRTStartup+0x157B12@0x559022[cataclysm-tiles.exe+0x159022]
    raise+0x200@0x7FFF3B6BAD70[msvcrt.dll+0x2AD70]
    mainCRTStartup+0x2740D1@0x6755E1[cataclysm-tiles.exe+0x2755E1]
    mainCRTStartup+0x2A24D0@0x6A39E0[cataclysm-tiles.exe+0x2A39E0]
    mainCRTStartup+0x285047@0x686557[cataclysm-tiles.exe+0x286557]
    reference_table+0x21CC79@0x140A8D9[cataclysm-tiles.exe+0x100A8D9]
    __tmainCRTStartup+0x23D@0x4013ED[cataclysm-tiles.exe+0x13ED]
    WinMainCRTStartup+0x1B@0x4014FB[cataclysm-tiles.exe+0x14FB]
    BaseThreadInitThunk+0x14@0x7FFF39888364[KERNEL32.DLL+0x8364]
    RtlUserThreadStart+0x21@0x7FFF3B8D70D1[ntdll.dll+0x670D1]
```

**Comment by neitsa** [comment]
Thank you very much, both of you, for your comments.

Thank you ZhilkinSerg for testing cv2pdb. It seems that there's might be something wrong that prevents the stack trace code to display function's names though...

Given your feedback I'm going to investigate with cv2pdb and clang 8.0 and then report any progress. At some point I should be able to do a PR about that.

Thanks!

**Comment by neitsa** [comment]
Tried the cv2pdb information, wondering why we couldn't get a meaningful stack trace with it. After looking at the cv2pdb  source code it appears it only understands `Dwarf-2` information (while gcc produces `Dwarf-4` symbols). 

C++ information are generated with DWARF 4 symbols (which is why the resulting PDB only contains C linkage functions which are basically emitted with dwarf 2 symbolic information). I don't intend to spend too much time on modifying cv2pdb to support dwarf4...

A possibility would be to force the gcc build chain for Windows to emit dwarf-2 through the  [`--gdwarf-2`](https://gcc.gnu.org/onlinedocs/gcc/Debugging-Options.html) switch. Haven't tried it yet as I need to setup a complete C:DDA build chain on my linux box.

One other thing is to compile the game source with clang 8.0 to generate MS symbolic information. I haven't been able to compile even a simple program with clang **and** the mingw-64 headers on a linux box... Maybe I did something wrong though (clang chokes on various gcc only stuff in the mingw headers). I'll try to address this. If I can't I'll need to test with the windows SDK headers installed on a linux box...

To be continued...

**Comment by ZhilkinSerg** [comment]
Just to make sure: did you follow [cross-compile instruction](https://github.com/CleverRaven/Cataclysm-DDA/blob/master/COMPILING.md#cross-compile-to-windows-from-linux)?

Also you can use docker image to make cross-compiling a bit simpler - https://github.com/ralreegorganon/cddadockerwin

**Comment by neitsa** [comment]
Yes. I just didn't have time lately... To be honest I didn't see the docker image though. Thanks for your answer and your help!

**Comment by neitsa** [comment]
# Update

## cv2pdb 

Compiled the game with mingw on a linux box and forced gcc to prodcue DWARF-2 symbols. Then used `cv2pdb.exe` **on Windows** to produce a pdb file from the dwarf symbolic information. The crash handler produce a meaningful stack trace!

Now there one big **major problem**: cv2pdb is in fact loading dynamically a lot of stuff under the hood from Visual Studio (I was able to confirm that by tracing the program with procmon).

After getting the required binaries (from Windows to Linux) I tried to start cv2pdb in wine but it doesn't work (there is no error output, it's simply stuck). Looking at the registry keys used by the program (more precisely by the pdb object generator from visual studio / cl) there are a lot of instantiated COM classes which are obviously not portable to wine... 

* End result: `cv2pdb` is a no go on a linux build chain.

## Crash dump

Another problem I noticed is that, for an unknown reason, the SEH handling of gcc is not exactly the same as the one from cl (MS compiler). The exception pointer obtained from a binaries compiled with gcc is not working correctly (it's basically empty) and the produced crash dump is exploitable...

It's quite strange because I'm quite sure, however, that the default exception handling for mingw-x64 is set to `seh` (another problem is that it seems to be `sjlj` for i686 which would add another problem to get a correct core dump on Windows because the system requires SEH).

## Next Step

Next step is to try clang / clang-cl (8.0). The only real problem here is that it probably requires the Windows headers and not the mingw one (I'm not really sure though. Last time I tried with a fresh install of llvm 8 from the llvm repo, clang choked on the mingw headers). 

I'll try to compile llvm with MXE and see what I can do with that... But first of all I have to make MXE understand that I want llvm 8.0 and not 7.0...

**Comment by neitsa** [comment]
# Update

After many days and many tries, I have been able to generate a fully functioning Windows executable and its full PDB from clang-cl; without having to rely on a Windows system (everything done on linux).

![pdb](https://user-images.githubusercontent.com/4538670/56504060-a5b0ee80-6517-11e9-930b-b57e70dc18aa.png)

The main difficulty is actually to gather all the headers and libraries without any VS installation and set them up in a fashion close to the one on a visual studio install so clang-cl can find everything it wants.

Windbg and visual studio (on Windows) can debug the executable - and use the symbols - without any problems.

I wrote an extensive how-to, which I plan to push on the repo when it will be ready.

## Next Step

The next step is to write a makefile and then compile the whole game.

**Comment by ZhilkinSerg** [comment]
Should be addressed by #36615.
